guides/custom_video.md

```diff
@@ -0,0 +1,215 @@
+
+The following snippet contains boilerplate for maintaining a list of sinks that have been added to the source; this allows all sources to be forked to multiple targets (e.g. transmission and local rendering). See [VideoContentHint](https://aws.github.io/amazon-chime-sdk-ios/Enums/VideoContentHint.html) for more information on the effects of that paramater to the downstream encoder.
+
+```
+class MyVideoSource: VideoSource {
+    // Do not indicate any hint to downstream encoder
+    var videoContentHint = VideoContentHint.none
+
+    // Downstream video sinks
+    private let sinks = NSMutableSet()
+
+    func startProducingFrames() {
+        while (true) {
+            // Obtain pixel buffer from undelying source ...
+
+            // Create frame
+            let buffer = VideoFramePixelBuffer(pixelBuffer: somePixelBuffer)
+            let timestampNs = someTimestamp
+            let frame = VideoFrame(timestampNs: Int64(timestampNs),
+                                   rotation: .rotation0,
+                                   buffer: buffer)
+
+            // Forward the frame to downstream sinks
+            for sink in sinks {
+                (sink as? VideoSink)?.onVideoFrameReceived(frame: frame)
+            }
+        }
+    }
+
+    func addVideoSink(sink: VideoSink) {
+        sinks.add(sink)
+    }
+
+    func removeVideoSink(sink: VideoSink) {
+        sinks.remove(sink)
+    }
+}
+```
+
+When enabling local video, call [AudioVideoControllerFacade.startLocalVideo(source:)](https://aws.github.io/amazon-chime-sdk-ios/Protocols/AudioVideoControllerFacade.html#/c:@M@AmazonChimeSDK@objc(pl)AudioVideoControllerFacade(im)startLocalVideoWithSource:) with the custom source as the parameter. Ensure that the capture source is started to start transmitting frames.
+
+```
+    // Create and start the processor
+    let myVideoSource = MyVideoSource()
+    myVideoSource.startProducingFrames()
+
+    // Begin transmission of frames
+    audioVideo.startLocalVideo(source: myVideoSource)
+```
+
+## Implementing a custom video processing step for local source
+
+By combining the [VideoSource](https://aws.github.io/amazon-chime-sdk-ios/Protocols/VideoSource.html) and [VideoSink](https://aws.github.io/amazon-chime-sdk-ios/Protocols/VideoSink.html) APIs, builders can easily create a video processing step to their applications. Incoming frames can be processed, and then fanned out to downstream sinks like in the following snippet. Note that if frames are passed onto seperate threads, builders must call [VideoFrame.retain](https://aws.github.io/amazon-chime-sdk-ios/) to avoid the resources being dropped before the seperate thread accesses them. See example processors in [Demo code](https://github.com/aws/amazon-chime-sdk-ios/blob/master/AmazonChimeSDKDemo/AmazonChimeSDKDemo/utils/MetalVideoProcessor.swift) for complete, documented implementations.
```

Comment:
There is no retain on `VideoFrame` API, and it's not in `MetailVideoProcessor` either. Is this needed for iOS?

Reply:
Woops, missed that (even had the dead link :P), thanks